### PR TITLE
fix(TCOMP-388): Update Beam API for all components.

### PR DESCRIPTION
**Pull request by RyanSkraba**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows Talend standard
- [X] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

We are using the Beam 0.3.0-incubating API which is stable, but broken for HDFS on Spark.

**What is the new behavior?**

We are using the Beam 0.4.0-incubating-SNAPSHOT API (soon to be released at 0.4.0).

**Does this PR introduce a breaking change?**

- [X] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

Upgrading the Beam API may have an effect on Beam execution runtimes.

**Other information**:

### CODE REVIEW

**Comment by RyanSkraba** [comment]
There was a lot of breakage in the HDFS components, due to some minor mistakes that broke when running in the DirectRunner (improved in 0.4.0-incubating).  They are also fixed in this PR.

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/287/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/287/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/295/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/295/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/296/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/296/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/297/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/297/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/298/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/298/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/299/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/299/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/301/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/301/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/302/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/302/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/303/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/303/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/307/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/307/)

**Comment by imejia** [comment]
CC: @jbonofre

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/310/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/310/)

File: core/components-adapter-beam-parent/pom.xml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    </parent>
+
+    <groupId>org.talend.components</groupId>
+    <artifactId>components-adapter-beam-parent</artifactId>
+    <packaging>pom</packaging>
+
+    <name>Component API - Beam Parent POM</name>
+
+    <properties>
+        <beam.version>0.4.0-incubating-SNAPSHOT</beam.version>
```
Comment (StefanLMNG) [inline core/components-adapter-beam-parent/pom.xml]: The beam version will be "0.4.0-incubating".

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/simplefileio/runtime/SimpleFileIoDatastoreRuntimeTest.java [hunk 2]
```diff
@@ -62,7 +62,7 @@ public void testRuntimeInfo() throws MalformedURLException {
         // We test the maven dependencies in this runtime class, where they are available.
         List<URL> dependencies = runtimeInfo.getMavenUrlDependencies();
         assertThat(dependencies, notNullValue());
-        assertThat(dependencies, hasItem(new URL("mvn:org.apache.beam/beam-sdks-java-io-hdfs/0.3.0-incubating/jar")));
+        assertThat(dependencies, hasItem(new URL("mvn:org.apache.beam/beam-sdks-java-io-hdfs/0.4.0-incubating-SNAPSHOT/jar")));
```
Comment (StefanLMNG) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/simplefileio/runtime/SimpleFileIoDatastoreRuntimeTest.java]: The beam version will be "0.4.0-incubating".

**Comment by RyanSkraba** [comment]
I'm taking a look at the build failures -- they're entirely within the `SimpleFileIoInputRuntimeTest`; everything else succeeds.

**Comment by RyanSkraba** [comment]
OK.  I've isolated the bug with the `SimpleFileIoInputRuntimeTest` -- I didn't manage to fix it, but I found why it used to work.  It's related to default target parallelism in the DirectRunner.  There's a TODO in the code, but it shouldn't stop this PR.

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        HDFSFileSource<LongWritable, Text> source = HDFSFileSource.from(fileSpec, TextInputFormat.class, LongWritable.class,
+                Text.class);
+
+        // Create a pipeline using the input component to get records.
+        DirectOptions options = PipelineOptionsFactory.as(DirectOptions.class);
+        options.setRunner(DirectRunner.class);
+        final Pipeline p = Pipeline.create(options);
+
+        Consumer<Text> consumer = new Consumer<Text>() {
+
```
Comment (imejia) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java]: remove space, yes, I am that annoying :P

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        // Create an input runtime based on the properties.
+        String fileSpec = mini.getFs().getUri().resolve("/user/test/").toString();
+        HDFSFileSource<LongWritable, Text> source = HDFSFileSource.from(fileSpec, TextInputFormat.class, LongWritable.class,
+                Text.class);
+
+        // Create a pipeline using the input component to get records.
+        DirectOptions options = PipelineOptionsFactory.as(DirectOptions.class);
+        options.setRunner(DirectRunner.class);
+        final Pipeline p = Pipeline.create(options);
```
Comment (imejia) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java]: Doesn't TestPipeline work here ?

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        RecordSetUtil.writeRandomAvroFile(mini.getFs(), "/user/test/input.avro", rs);
+        String fileSpec = mini.getFs().getUri().resolve("/user/test/input.avro").toString();
+
+        // Create an input runtime based on the properties.
+        AvroHDFSFileSource<IndexedRecord> source = new AvroHDFSFileSource(fileSpec, AvroCoder.of(rs.getSchema()));
+
+        // Create a pipeline using the input component to get records.
+        DirectOptions options = PipelineOptionsFactory.as(DirectOptions.class);
+        options.setRunner(DirectRunner.class);
+        final Pipeline p = Pipeline.create(options);
```
Comment (imejia) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java]: Same for TestPipeline here (it auto configures the direct runner).

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/RecordSetUtil.java [hunk 6]
```diff
@@ -106,4 +106,30 @@ public static void writeRandomAvroFile(FileSystem fs, String path, RecordSet td)
...
+        try (PrintWriter w = new PrintWriter(fs.create(new Path(path)))) {
+            for (List<IndexedRecord> partition : td.getPartitions()) {
+                for (IndexedRecord record : partition) {
+                    if (size > 0)
+                        w.print(record.get(0));
+                    for (int i = 1; i < size; i++) {
+                        w.print(';');
+                        w.print(record.get(i));
+                    }
+                    w.print('\n');
```
Comment (imejia) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/RecordSetUtil.java]: System.getProperty("line.separator") ? what about windows users !! think about those poor souls.

File: core/components-adapter-beam-parent/pom.xml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
+            <version>1.3</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.beam</groupId>
+            <artifactId>beam-runners-direct-java</artifactId>
+            <version>${beam.version}</version>
```
Comment (imejia) [inline core/components-adapter-beam-parent/pom.xml]: omit version (already in dependencyManagement), I wonder if you can omit the whole dependency) or even better the whole dependencyManagement:
Selon JB:
1. soit pas de dependances dans le parent et le définir dans les sub
2. soit tu mets un dependencies dans le parent, mais alors pas de dependencyManagement (pas de sens puisque ça vient déjà du parent)

**Comment by RyanSkraba** [comment]
Hello!  Forgive me for the scope change -- I couldn't see any reason not to update the JDBC component family at the same time.  There's a TODO in JDBC with respect to manually setting the Beam API version in the pom, but I expect there to be a later PR splitting the Beam and DI runtimes.

File: core/components-adapter-beam/src/test/java/org/talend/components/adapter/beam/transform/DirectConsumerCollectorTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        Consume<Integer> afterConsumer = new Consume<>();
+
+        // Creating the collectors in the try-with-resources ensures that they are automatically cleaned up. Otherwise,
+        // the close() method on each instance should be called after the test.
+        try (DirectConsumerCollector<String> before = DirectConsumerCollector.of(beforeConsumer); //
+                DirectConsumerCollector<Integer> after = DirectConsumerCollector.of(afterConsumer);) {
+
+            PipelineOptions options = PipelineOptionsFactory.create();
+            options.setRunner(DirectRunner.class);
+            final Pipeline p = Pipeline.create(options);
```
Comment (imejia) [inline core/components-adapter-beam/src/test/java/org/talend/components/adapter/beam/transform/DirectConsumerCollectorTest.java]: same about TestPipeline

File: core/components-adapter-beam/src/test/java/org/talend/components/adapter/beam/transform/DirectConsumerCollectorTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            PCollection<Integer> output = input.apply(ParDo.of(new DirectCollectorTest.GetLength()));
+            output.apply(after);
+
+            // Optional -- validate the contents of the collections in the pipeline.
+            PAssert.that(input).containsInAnyOrder("one", "two", "three");
+            PAssert.that(output).containsInAnyOrder(3, 3, 5);
+
+            // Run the pipeline to fill the collectors.
+            PipelineResult pr = p.run();
+            DirectRunner.DirectPipelineResult dpr = (DirectRunner.DirectPipelineResult) pr;
```
Comment (imejia) [inline core/components-adapter-beam/src/test/java/org/talend/components/adapter/beam/transform/DirectConsumerCollectorTest.java]: unused and why cast to the DirectPipelineResult one

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/319/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/319/)

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/simplefileio/runtime/SimpleFileIoDatastoreRuntimeTest.java (hunk 2 above)
Comment (RyanSkraba) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/simplefileio/runtime/SimpleFileIoDatastoreRuntimeTest.java]: Doesn't exist yet!  I'd prefer to remove the SNAPSHOT in a second PR -- we'll see who wins first (the 0.4.0 vote or merge approval on this PR!)

File: components/components-fileio/simplefileio-runtime/src/main/java/org/talend/components/simplefileio/runtime/SimpleFileIoInputRuntime.java [hunk 10]
```diff
@@ -80,18 +84,22 @@ public ValidationResult initialize(RuntimeContainer container, SimpleFileIoInput
...
-                    .apply((PTransform) ConvertToIndexedRecord.of());
+
+            PCollection<KV<org.apache.hadoop.io.LongWritable, Text>> pc1 = in.apply(Read.from(source));
+
+            PCollection<Text> pc2 = pc1.apply(Values.<Text> create());
+
+            PCollection<String[]> pc3 = pc2.apply(ParDo.of(new ExtractCsvSplit(
+                    properties.datasetRef.getReference().fieldDelimiter.getValue())));
+
+            PCollection pc4 = pc3.apply(ConvertToIndexedRecord.<String[], IndexedRecord> of());
```
Comment (RyanSkraba) [inline components/components-fileio/simplefileio-runtime/src/main/java/org/talend/components/simplefileio/runtime/SimpleFileIoInputRuntime.java]: Hello!  Thanks for the review!

I had quite a few problems during pipeline construction, and separating them makes it easier to debug (and not necessarily less readable imo).

The raw PCollection is unavoidable for the moment, while we're figuring out the LazyAvroCoder strategy...

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java (hunk 3 above)
Comment (RyanSkraba) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java]: talend-java-formatter says no!

File: core/components-adapter-beam/src/test/java/org/talend/components/adapter/beam/transform/DirectConsumerCollectorTest.java (hunk 9 above)
Comment (RyanSkraba) [inline core/components-adapter-beam/src/test/java/org/talend/components/adapter/beam/transform/DirectConsumerCollectorTest.java]: Done!

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java (hunk 5 above)
Comment (RyanSkraba) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java]: Done!

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/RecordSetUtil.java (hunk 6 above)
Comment (RyanSkraba) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/RecordSetUtil.java]: Done!  Note that a lot of this RecordSet and stuff will probably be re-homed to a `components-adapter-beam-testing` module eventually.

File: components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java (hunk 4 above)
Comment (RyanSkraba) [inline components/components-fileio/simplefileio-runtime/src/test/java/org/talend/components/test/MiniDfsResourceTest.java]: Done!

**Comment by RyanSkraba** [comment]
This PR is still at 0.4.0-incubating-SNAPSHOT.  We'll do another PR when the release version is available -- it should be *much* simpler.

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/332/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/332/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/349/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/349/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/352/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/352/)

**Pull request merged**